File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Palindrome Partitioning/Java/Solution.java
import java.util.*;

class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        String[] sArr = new String[s.length()];
        for (int i = 0; i < s.length(); i++) {
            sArr[i] = String.valueOf(s.charAt(i));
        }
        backTracking(sArr, result, 0, new ArrayList<>());
        return result;
    }

    private void backTracking(String[] sArr, List<List<String>> result, int start, List<String> subStr) {
        if (start == sArr.length) {
            if (isPalindrome(subStr)) {
                result.add(new ArrayList<>(subStr));
            }
            return;
        }

        String s = "";
        for (int i = start; i < sArr.length; i++) {
            s += sArr[i];
            subStr.add(s);
            backTracking(sArr, result, i + 1, subStr);
            subStr.remove(subStr.size() - 1);
        }
    }

    private boolean isPalindrome(List<String> subStr) {
        for (String str : subStr) {
            if (!str.equals(new StringBuilder(str).reverse().toString())) {
                return false;
            }
        }

        return true;
    }
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/Palindrome Partitioning/solution.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        # 메서드
        def backTracking(answer, candi, s):
            if s == "":
                if isPalindrome(candi):
                    answer.append(candi[:])
                return

            for i in range(len(s)):
                candi.append(s[:i + 1])
                backTracking(answer, candi, s[i + 1:])
                candi.pop()
            
        def isPalindrome(candi):
            for c in candi:
                if c != c[::-1]:
                    return False
            
            return True

        # 메인 로직
        answer = []
        backTracking(answer, [], s)
        return answer
